### PR TITLE
Update Durable Nonces Course

### PR DESCRIPTION
**Pull request by darkvallen**
### Summary of Changes
- Updated code snippets
- Use Partial signing for transactions with multiple signers.
- Use `getExplorerLink` from solana-helpers for easier link generation.
- Added the calculation for the exact rent-exempt balance using `connection.getMinimumBalanceForRentExemption`.
- Added screenshots of successful test runs are included in the documentation.
- Fixed grammar and styling
- Fixed as per guidelines

Fixes #
Unboxed PRs
https://github.com/Unboxed-Software/solana-lab-durable-nonces/pull/1

### CODE REVIEW

File: content/courses/offline-transactions/durable-nonces.md [hunk 1]
```diff
@@ -14,39 +14,39 @@ description: "Use durable nonces to sign transactions ahead of time."
...
    program. These are executed in order.
 
-2. **Signatures**: Signatures are the proof that the transaction was signed by
-   the required singers/authorities. For instance, if you are transferring SOL
-   from your wallet to another, you'll need to sign the transaction so the
-   network can verify that the transaction is valid.
+2. **Signatures**: Signatures are proof that the transaction was signed by the
+   required signers/authorities. For instance, if you are transferring SOL from
+   your wallet to another, you'll need to sign the transaction so the network
+   can verify that the transaction is valid.
```
Comment (mikemaccana) [inline content/courses/offline-transactions/durable-nonces.md:41]: This is a criticism of the old writing, but the logic here is circular and bad. Signatures prove I sign it, which means I have a signature, which proves I signed it. 

Try:

> **Signatures**: Signatures are proof that the transaction was made by the required signers (assuming they have kept their secret keys secure).

File: content/courses/offline-transactions/durable-nonces.md [hunk 2]
```diff
@@ -14,39 +14,39 @@ description: "Use durable nonces to sign transactions ahead of time."
...
+  instruction, the nonce will still be advanced, even though all other
+  instructions will be reverted.
 
 ## Overview
 
 Durable Nonces are a way to bypass the expiration date of regular transactions.
-To understand that better, we'll start by looking at the concepts behind regular
-transactions.
+To understand this more clearly, we'll start by looking at the concepts behind
+regular transactions.
```
Comment (mikemaccana) [inline content/courses/offline-transactions/durable-nonces.md:30]:
```suggestion
To understand this, we'll start by looking at the concepts behind
regular transactions.
```

Comment (mikemaccana) [reply content/courses/offline-transactions/durable-nonces.md:30]: Don't be afraid to clean up unnecessarily wordy language.

File: content/courses/offline-transactions/durable-nonces.md [hunk 3]
```diff
@@ -14,39 +14,39 @@ description: "Use durable nonces to sign transactions ahead of time."
...
 
-In Solana, transactions are made of three main parts:
+In Solana, transactions consist of three main parts:
```
Comment (mikemaccana) [inline content/courses/offline-transactions/durable-nonces.md:32]:
```suggestion
In Solana, transactions have three main parts:
```

Plain language.

File: content/courses/offline-transactions/durable-nonces.md [hunk 4]
```diff
@@ -115,30 +115,30 @@ From the
...
+transaction signs away 500 SOL to the attacker and changes the nonce authority
+to the attacker as well. Let's say you don't have this amount yet, but in the
+future, you will. This is insidious, as the attacker would wait to cash this
+check as soon as your balance exceeeds 500 SOL. And you'll have no recollection
+of what you clicked on. It can remain dormant for days, weeks, or years.
 
-This is not meant to provoke hysteria, just as a PSA of what's possible. This is
-why you should only put into hot wallets what you're willing to lose and don't
-sign with your cold wallet.
+This is not meant to provoke hysteria, but serves as a PSA of what's possible.
```
Comment (mikemaccana) [inline content/courses/offline-transactions/durable-nonces.md:128]:
```suggestion
This is not meant to provoke hysteria, but serves to show what's possible.
```

PSAs are US-specific.

Comment (mikemaccana) [reply content/courses/offline-transactions/durable-nonces.md:128]: This lesson goes from 'you are the developer' to 'you are the user'. We should fix that.

Maybe as developers, we tell people their durable nonce trandsactions may be flagged by wallets, and then explain that durable nonces are often used for malicious transactions?

File: content/courses/offline-transactions/durable-nonces.md [hunk 5]
```diff
@@ -115,30 +115,30 @@ From the
...
+This is why you should only keep in hot wallets what you're willing to lose and
+avoid signing with your cold wallet.
 
-### Using Durable nonces overcome the short lifespan of regular transactions
+### Using Durable Nonces to Overcome the Short Lifespan of Regular Transactions
 
-Durable nonces are a way to sign transactions offchain and keep them in storage
-until they are ready to be submitted to the network. And this allows us to
-create durable transactions.
+Durable nonces are a way to sign transactions off-chain and keep them in storage
```
Comment (mikemaccana) [inline content/courses/offline-transactions/durable-nonces.md:134]: offchain is one word, please fix everywhere.

File: content/courses/offline-transactions/durable-nonces.md [hunk 6]
```diff
@@ -150,72 +150,75 @@ It is important to note that durable nonces require
 to function, thus they have some special rules that don't apply normally. We'll
 see this as we deep dive into the technicals.
 
-### Durable nonces in-depth
+### Durable Nonces In-Depth
 
 Durable transactions differ from regular transactions in the following ways:
 
 1. Durable Nonces replace the recent blockhash with a nonce. This nonce is
    stored in a `nonce account` and will be used only once in one transaction.
```
Comment (mikemaccana) [inline content/courses/offline-transactions/durable-nonces.md:158]: Not sure `nonce account` is code or a file name. Maybe just 'nonce account' ?

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Great work, a few minor bits but it's very clear you've read the contributing guidelines. Much appreciated, let's get this merged!

File: content/courses/offline-transactions/durable-nonces.md (hunk 4 above)
Comment (darkvallen) [reply content/courses/offline-transactions/durable-nonces.md]: I have revised the section to maintain a consistent perspective, addressing both developers and users. The updated text now informs developers to educate users about the potential risks and how to mitigate them.

Please let me know if there are any further adjustments needed.

**Comment by darkvallen** [comment]
> Great work, a few minor bits but it's very clear you've read the contributing guidelines. Much appreciated, let's get this merged!

Updated, Could you please review the updated "Considerations" section and let me know if it addresses your concerns?

**Comment by mikemaccana** [comment]
Marked as winner, congrats!

**Comment by mikemaccana** [comment]
Also @darkvallen: for https://github.com/Unboxed-Software/solana-lab-durable-nonces can you please make your changes to the `solution` branch, not `new-solution`? Also please update `starter` branch if necessary.

**Comment by darkvallen** [comment]
> Also @darkvallen: for https://github.com/Unboxed-Software/solana-lab-durable-nonces can you please make your changes to the `solution` branch, not `new-solution`? Also please update `starter` branch if necessary.
> 
> Reply here once done.

Thanks @mikemaccana! A new learning experience for me.

Done https://github.com/Unboxed-Software/solana-lab-durable-nonces/pull/4 here's the `solution` branch PR.

**Pull request merged**